lib/configuration.js

```diff
@@ -15,150 +15,430 @@
-
-    if (has(initConfiguration.serviceContext, 'service')) {
-
-      defaultReturnObject.service = initConfiguration.serviceContext.service;
+  var gaeModuleVer = env.GAE_MODULE_VERSION;
+  this._serviceContext.service = isString(gaeModuleName) ? gaeModuleName : '';
+  this._serviceContext.version = isString(gaeModuleVer) ? gaeModuleVer : '';
+  if (isPlainObject(this._givenConfiguration)
+    && isPlainObject(this._givenConfiguration.serviceContext)) {
+    if (isString(this._givenConfiguration.serviceContext.service)) {
+      this._serviceContext.service = 
+        this._givenConfiguration.serviceContext.service;
     }
-
-    if (has(initConfiguration.serviceContext, 'version')) {
-
-      defaultReturnObject.version = initConfiguration.serviceContext.version;
+    if (isString(this._givenConfiguration.serviceContext.version)) {
+      this._serviceContext.version =
+        this._givenConfiguration.serviceContext.version;
     }
   }
-
-  return defaultReturnObject;
-}
-
+};
 /**
- * @typedef NormalizedConfigurationVariables
- * @type Object
- * @property {String} projectId - the project id of the hosting application
- * @property {Boolean} reportUncaughtExceptions=true - the flag denoting
- *  whether or not to catch and report uncaught errors to the API
- * @property {Object} serviceContext - the hosting applications service
- *  information
- * @property {String} serviceContext.service - the hosting applications service
- *  that it is running atop of
- * @property {String} serviceContext.version - the hosting applications version
- * @property {String} key - an API key used to communicate with the service
- * @property {String} version - the version of the module
- * @property {Boolean} shouldReportErrorsToAPI - a flag indicating whether or
- *  not the Middleware should send errors to the Stackdriver Error Reporting
- *  API. This flag is derived from the `NODE_ENV` environmental varaible being
- *  set to production: if the flag is set to 'production' then the
- *  `shouldReportErrorsToAPI` property will be set to true and the API client
- *  will attempt to send errors to the Stackdriver Error Reporting API.
+ * The _gatherLocalConfiguration function is responsible for determining
+ * directly determing whether the properties `reportUncaughtExceptions` and
+ * `key`, which can be optionally supplied in the runtime configuration, should
+ * be merged into the instance. This function also calls several specialized
+ * environmental variable checkers which not only check for the optional runtime
+ * configuration supplied values but also the processes environmental values.
+ * @memberof Configuration
+ * @private
+ * @function _gatherLocalConfiguration
+ * @returns {Undefined} - does not return anything
  */
-
+Configuration.prototype._gatherLocalConfiguration = function() {
+  if (isPlainObject(this._givenConfiguration)) {
+    if (isBoolean(this._givenConfiguration.reportUncaughtExceptions)) {
```

Comment:
Should we silently ignore things uncaught exceptions if the user accidentally sets this to '1' instead of `true`? Either throw or be more resilient to bad user input and do the right thing.